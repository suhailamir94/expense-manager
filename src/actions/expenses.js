export const addExpense = (
  { description = "", note = "", amount = 0, createdAt = 0 },
  userId
) => ({
  type: "ADD_EXPENSE",
  expense: {
    description,
    note,
    amount,
    createdAt,
  },
  userId,
});

//EDIT_EXPENSE
export const removeExpense = ({ id, userId }) => ({
  type: "REMOVE_EXPENSE",
  id,
  userId,
});

//REMOVE_EXPENSE
export const editExpense = (id, updates, userId) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
  userId,
});
