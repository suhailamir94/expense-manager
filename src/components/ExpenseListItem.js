import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
    <h3>
      <Link to={`/edit/${id}`}>{description}</Link>
    </h3>
    <p>
      &#x20B9;{amount} - {moment(createdAt).format("MMMM Do, YYYY")}
    </p>
  </div>
);

export default ExpenseListItem;
