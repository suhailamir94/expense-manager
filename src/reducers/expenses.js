const expensesReducerDefaultState = [];
const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "GET_EXPENSES_SUCCEEDED":
      return [...action.expenses];
    case "ADD_EXPENSE_SUCCEEDED":
      return [...state, action.expense];
    case "REMOVE_EXPENSE_SUCCEEDED":
      return state.filter(e => e.id != action.id);
    case "EDIT_EXPENSE_SUCCEEDED":
      return state.map(expense =>
        expense.id === action.id ? { ...expense, ...action.updates } : expense
      );
    case "ADD_EXPENSE_FAILED":
      return state;
    case "GET_EXPENSES_FAILED":
      return state;
    case "EDIT_EXPENSE_FAILED":
      return state;
    case "REMOVE_EXPENSE_FAILED":
      return state;
    default:
      return state;
  }
};

export default expensesReducer;
