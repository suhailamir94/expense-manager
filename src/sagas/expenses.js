import { call, put, takeEvery, takeLatest } from "redux-saga/effects";
import {
  addExpenseFirebase,
  getAllExpensesFirebase,
  editExpenseFirebase,
  removeExpenseFirebase,
} from "../firebase/firebase";

// worker Saga: will be fired on USER_FETCH_REQUESTED actions
function* addExpenseAsync(action) {
  try {
    const id = yield call(addExpenseFirebase, [action.expense, action.userId]);
    yield put({
      type: "ADD_EXPENSE_SUCCEEDED",
      expense: { id, ...action.expense },
    });
  } catch (e) {
    yield put({ type: "ADD_EXPENSE_FAILED", message: e.message });
  }
}
function* editExpenseAsync({ id, updates, userId }) {
  try {
    yield call(editExpenseFirebase, [id, updates, userId]);
    yield put({
      type: "EDIT_EXPENSE_SUCCEEDED",
      id,
      updates,
    });
  } catch (e) {
    yield put({ type: "ADD_EXPENSE_FAILED", message: e.message });
  }
}
function* removeExpenseAsync({ id, userId }) {
  try {
    yield call(removeExpenseFirebase, [id, userId]);
    yield put({
      type: "REMOVE_EXPENSE_SUCCEEDED",
      id,
    });
  } catch (e) {
    yield put({ type: "REMOVE_EXPENSE_FAILED", message: e.message });
  }
}
function* getExpensesAsync(action) {
  try {
    const expenses = yield call(getAllExpensesFirebase, action.userId);
    yield put({
      type: "GET_EXPENSES_SUCCEEDED",
      expenses,
    });
  } catch (e) {
    yield put({ type: "GET_EXPENSES_FAILED", message: e.message });
  }
}
/*
  Starts fetchUser on each dispatched `USER_FETCH_REQUESTED` action.
  Allows concurrent fetches of user.
*/

export function* getAllExpensesSaga() {
  yield takeEvery("GET_EXPENSES", getExpensesAsync);
}

export function* addExpenseSaga() {
  yield takeEvery("ADD_EXPENSE", addExpenseAsync);
}

export function* editExpenseSaga() {
  yield takeEvery("EDIT_EXPENSE", editExpenseAsync);
}

export function* removeExpenseSaga() {
  yield takeEvery("REMOVE_EXPENSE", removeExpenseAsync);
}
