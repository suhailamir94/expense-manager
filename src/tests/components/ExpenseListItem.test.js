import React from "react";
import { shallow } from "enzyme";
import ExpenseListItem from "../../components/ExpenseListItem";
import expenses from "../fixtures/expenses";

test("should render Expense List Item with expense", () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]} />);
  expect(wrapper).toMatchSnapshot();
});
