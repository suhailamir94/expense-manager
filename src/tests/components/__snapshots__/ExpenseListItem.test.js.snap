// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense List Item with expense 1`] = `
<div>
  <h3>
    <Link
      to="/edit/1"
    >
      gum
    </Link>
  </h3>
  <p>
    195
     - 
    0
  </p>
</div>
`;
